atlasdb-timelock-server/src/test/java/com/palantir/atlasdb/timelock/paxos/PaxosTimestampBoundStoreTest.java

```diff
@@ -0,0 +1,324 @@
+        assertThat(sequenceAndBound.getBound()).isEqualTo(0);
+    }
+
+    @Test
+    public void canGetAgreedState() {
+        store.storeUpperLimit(TIMESTAMP_1);
+        PaxosTimestampBoundStore.SequenceAndBound sequenceAndBound = store.getAgreedState(1);
+        assertThat(sequenceAndBound.getSeqId()).isEqualTo(1);
+        assertThat(sequenceAndBound.getBound()).isEqualTo(TIMESTAMP_1);
+    }
+
+    @Test
+    public void canSafelyGetAgreedStateFromPrehistory() {
+        assertThat(store.getAgreedState(Long.MIN_VALUE).getBound()).isEqualTo(0);
+    }
+
+    @Test
+    public void canGetAgreedStateAfterNodeDown() {
+        int nodeId = 1;
+        failureToggles.get(nodeId).set(true);
+        store.storeUpperLimit(TIMESTAMP_1);
+        PaxosTimestampBoundStore additionalStore = createPaxosTimestampBoundStore(nodeId);
+        failureToggles.get(nodeId).set(false);
+
+        assertThat(additionalStore.getAgreedState(2).getBound()).isEqualTo(TIMESTAMP_1);
+    }
+
+    @Test
+    public void cannotGetAgreedStateFromTheFuture() {
+        assertThatThrownBy(() -> store.getAgreedState(Long.MAX_VALUE))
+                .isInstanceOf(NullPointerException.class);
+    }
+
+    @Test
+    public void canSafelyForceAgreedStateFromPrehistory() {
+        assertThat(store.forceAgreedState(Long.MIN_VALUE, Long.MIN_VALUE).getBound()).isEqualTo(0);
+    }
+
+    @Test
+    public void canForceAgreedState() {
+        assertThat(store.forceAgreedState(1, FORTY_TWO)).isEqualTo(ONE_AND_FORTY_TWO);
+        assertThat(store.getAgreedState(1)).isEqualTo(ONE_AND_FORTY_TWO);
+    }
+
+    @Test
+    public void forceAgreedStateCanBeUsedToGainKnowledge() {
+        assertThat(store.forceAgreedState(1, FORTY_TWO)).isEqualTo(ONE_AND_FORTY_TWO);
+
+        PaxosTimestampBoundStore additionalStore = createPaxosTimestampBoundStore(1);
+        assertThat(additionalStore.forceAgreedState(1, null)).isEqualTo(ONE_AND_FORTY_TWO);
+    }
+
+    @Test
+    public void forceAgreedStateReturnsFirstForcedValue() {
+        assertThat(store.forceAgreedState(1, FORTY_TWO)).isEqualTo(ONE_AND_FORTY_TWO);
+        assertThat(store.forceAgreedState(1, 1L)).isEqualTo(ONE_AND_FORTY_TWO);
+        assertThat(store.getAgreedState(1)).isEqualTo(ONE_AND_FORTY_TWO);
+    }
+
+    @Test
+    public void forceAgreedStateOperatesAtSequenceNumberLevel() {
+        long fortyThree = FORTY_TWO + 1;
+        assertThat(store.forceAgreedState(1, FORTY_TWO)).isEqualTo(ONE_AND_FORTY_TWO);
+        assertThat(store.forceAgreedState(0, fortyThree)).isEqualTo(ImmutableSequenceAndBound.of(0, fortyThree));
+        assertThat(store.getAgreedState(0)).isEqualTo(ImmutableSequenceAndBound.of(0, fortyThree));
+        assertThat(store.getAgreedState(1)).isEqualTo(ONE_AND_FORTY_TWO);
+    }
+
+    @Test
+    public void forceAgreedStateThrowsIfNoStateWasAgreedUpon() {
+        assertThatThrownBy(() -> store.forceAgreedState(1, null)).isInstanceOf(NullPointerException.class);
```

Comment:
Hmm, the original code actually has an explicit `Preconditions.checkNotNull()` to throw this. The docs of NPE say that "Applications should throw instances of this class to indicate other illegal uses of the null object". I'm quite neutral on this (calling force with a null when it isn't agreed is an illegal use, so I think NPE is okay) - what do you think?